ext/standard/basic_functions.stub.php

```diff
@@ -1510,3 +1512,18 @@ function sapi_windows_set_ctrl_handler(?callable $handler, bool $add = true): bo
 
 function sapi_windows_generate_ctrl_event(int $event, int $pid = 0): bool {}
 #endif
+} // end global namespace
+
+namespace iterable {
+
+function any(iterable $iterable, ?callable $callback = null): bool {}
+
+function all(iterable $iterable, ?callable $callback = null): bool {}
+
+function none(iterable $iterable, ?callable $callback = null): bool {}
+
+function reduce(iterable $iterable, callable $callback, mixed $initial = null): mixed {}
```

Comment:
I strongly prefer removing `$initial` and adding `fold` which always requires it:

```php
function fold(iterable $iterable, callable $callback, mixed $initial): mixed {}
function reduce(iterable $iterable, callable $callback): mixed {}
```

The `fold` function doesn't throw on empty, and the `reduce` will. This pattern exists in other languages, such as Kotlin.

I'm happy to do this work if you'll agree.

Reply:
I feel like the inconsistency with array_reduce (which has optional $initial=null) would have more objectors for making it harder to learn the language or switch code from array_reduce to `*reduce` intuitively for beginners.

Reply:
There is an error condition in `reduce` where there is not an initial value, and it should throw. We should not repeat the mistakes of the past. You argue in another comment that `find` is useful in other languages, and yet you don't buy that same argument here? What gives?

Reply:
Looking at this again, I'd agree my earlier proposal for reduce was a mistake and it's worth changing, adding fold and either removing reduce entirely or requiring a non-empty array.

The other argument was about including a function, not a change

Reply:
I can work on `fold` tonight.

Reply:
Are you okay with these signatures and semantics for `fold` and `reduce`?

I imagine there is some discussion to be had for naming the parameters to make sure named parameters is a good experience, so let me know what you think. The name `$into` I [picked from Swift](https://developer.apple.com/documentation/swift/array/3126956-reduce). I used `$by` because it's short but not an abbreviation like `acc`; my quick glance in other languages' docs didn't turn up anything better.

```php
namespace Iterable;

/**
 * @template Element
 * @template Result
 * @param iterable<Element> $seq
 * @param Result $into
 * @param callable(Result, Element): Result $by
 * @return Result
 */
function fold(iterable $seq, mixed $into, callable $by): mixed {
    foreach ($seq as $value) {
        $into = $by($into, $value);
    }
    return $into;
}

/** Private helper, wouldn't actually be exposed.
 * @template Key
 * @template Value
 * @param iterable<Key, Value> $seq
 * @return \Iterator<Key, Value>
 */
function to_iterator(iterable $seq): \Iterator {
    if (\is_array($seq)) {
        return new \ArrayIterator($seq);
    } elseif ($seq instanceof \Iterator) {
        return $seq;
    } else {
        assert($seq instanceof \IteratorAggregate);
        return namespace\to_iterator($seq->getIterator());
    }
}


/**
 * @template Element
 * @param iterable<Element> $seq
 * @param callable(Element, Element): Element $by
 * @return Element
 * @throws \ValueError if $seq does not have at least 1 element
 */
function reduce(iterable $seq, callable $by): mixed {
    $iterator = namespace\to_iterator($seq);
    $iterator->rewind();
    if (!$iterator->valid()) {
        throw new \ValueError("parameter \$seq to reduce... was empty");
    }

    $into = $iterator->current();
    for ($iterator->next(); $iterator->valid(); $iterator->next()) {
        $into = $by($into, $iterator->current());
    }
    return $into;
}
```

Reply:
For fold, I think having the callback third would be hard to remember when it's second in other reduce() function

The inner implementation seem reasonable enough. I assume the ArrayObject is just for illustrating the behavior and not the internal implementation

$seq seems like an harder to remember naming choice compared to $array/$iterable used elsewhere - https://www.php.net/manual/en/function.iterator-apply.php and https://www.php.net/manual/en/function.array-walk.php - especially for non-english speakers 

PHP's already using $initial for https://www.php.net/manual/en/function.array-reduce.php and I don't see a strong reason to switch to a different name - initial's been used elsewhere (e.g. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce)

Reply:
> The inner implementation seem reasonable enough. I assume the ArrayObject is just for illustrating the behavior and not the internal implementation

The `ArrayIterator` is just for showing behavior, yes. Notably, this will _not_ pass `NULL` as the first parameter to the callback on the very first time it is called, unlike `array_reduce` and what this PR currently does.

As an example with the data set `[1, 3, 5, 7]`:

```php
$result = reduce([1, 3, 5, 7], function ($into, $value) {
    $retval = $into + $value;
    echo "fn ({$into}, {$value}) => {$retval}\n";
    return $retval;
});
```

This will print:

```
fn (1, 3) => 4
fn (4, 5) => 9
fn (9, 7) => 16
```

And not:

```
fn(, 1) => 1
fn (1, 3) => 4
fn (4, 5) => 9
fn (9, 7) => 16
```

Reply:
I implemented the changes to `reduce` on this branch: https://github.com/morrisonlevi/php-src/tree/levi/any-all-iterable-checks. For some reason it wouldn't let me select your fork as the merge-base, so I didn't open a PR, but you can look at the last two commits.